### ISSUE TITLE
Support wildcards in queries

### ISSUE THREAD

**Issue by shangyian** [DataJunction/dj#648]
Currently if we do something like:

```
SELECT * FROM node_a
```

DJ won't infer with all the columns from node_a and will instead provide a `WildcardType` column type, which isn't quite right.   Instead, the syntax above should result in in DJ populating the AST with all the columns from the projection that's being selected from so that type inference will just result in those column types.

This is largely for ease-of-use, as most users would prefer to write `SELECT *` rather than `SELECT <insert 200 columns>`, especially in a highly nested query.

**Comment by samredai** [comment]
How would we keep the node's static column metadata up to date as `node_a` changes? Would we have to search all downstream nodes that use `*` and update their columns anytime a node changes? Or are you thinking we wouldn't store any column metadata if the projection is just a wildcard and try and expand it at runtime? Would we also have to support wildcards within deeply nested subqueries?

**Comment by shangyian** [comment]
> Or are you thinking we wouldn't store any column metadata if the projection is just a wildcard and try and expand it at runtime?

Yeah, in thinking about this, it would be better if it gets expanded right when someone creates or updates the node, so the wildcard in SQL would immediately get replaced by the columns. Then users can go and remove or change the columns if needed if anything changes.

> Would we also have to support wildcards within deeply nested subqueries?

This is actually the exact use case I was coming across. I had five nested subqueries, each of which were querying wildcards on their respective projections, and I didn't like the idea of manually copying and pasting all the columns from each projection.

**Comment by shangyian** [comment]
@samredai in thinking about this more, I think it might be better if we always make sure that the update logic when a query for a given node changes will accurately update any downstream nodes + column metadata based on that. Otherwise a query that may otherwise be short (i.e., just `SELECT *`) will become extremely long.

**Comment by CircArgs** [comment]
@shangyian What's meant by "isn't quite right"? The implications seem to be that this is a compile issue and when a node with a wildcard is requested it references the table and injects the columns from that table (the table being compiled will have given us the columns at that stage). This sounds viable right?

**Comment by shangyian** [comment]
> What's meant by "isn't quite right"?

@CircArgs I was referring specifically to the column type we use, which is `WildcardType`. In my mind `WildcardType` means that it can be a single column of any type, not that it can be an arbitrary number of different columns, so I don't think we're using it the right way.

I agree that this is a compile issue and we should dynamically inject in all the columns when we see a wildcard. I think we should continue to save the query with the wildcard, and then if there's any updates to a parent node, we should also update all the downstream's column references.
